File: JavaScript solutions/2816. Double a Number Represented as a Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var doubleIt = function (head) {
  if (head === null) return head;

  let dummy = new ListNode();
  let current = dummy;
  let carry = 0;

  let prev = null;
  let temp = head;

  while (temp !== null) {
    let nextTemp = temp.next;
    temp.next = prev;

    prev = temp;
    temp = nextTemp;
  }

  while (prev !== null) {
    let sum = carry;
    sum += prev.val + prev.val;
    prev = prev.next;

    carry = Math.floor(sum / 10);
    current.next = new ListNode(sum % 10);
    current = current.next;
  }

  if (carry > 0) {
    current.next = new ListNode(carry);
  }

  prev = null;
  temp = dummy.next;

  while (temp !== null) {
    let nextTemp = temp.next;
    temp.next = prev;

    prev = temp;
    temp = nextTemp;
  }

  return prev;
};

File: C++ solutions/2816. Double a Number Represented as a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* doubleIt(ListNode* head) {
        if(head==nullptr) return head;
        
        ListNode* dummy = new ListNode();
        ListNode* current = dummy;
        int carry = 0;
        
        ListNode* prev = nullptr;
        ListNode* temp = head;
        
        while (temp != nullptr) {
            ListNode* nextTemp = temp->next; 
            temp->next = prev;
            
            prev = temp;
            temp = nextTemp;
        }
        
       
        
        while (prev != nullptr) {
            int sum = carry;
            sum += (prev->val + prev->val);
            prev = prev->next;
           
            
            carry = sum / 10;
            current->next = new ListNode(sum % 10);
            current = current->next;
        }
        
        if (carry > 0) {
            current->next = new ListNode(carry);
        }
        
        prev = nullptr;
        temp = dummy->next;
        
        while (temp != nullptr) {
            ListNode* nextTemp = temp->next; 
            temp->next = prev;
            
            prev = temp;
            temp = nextTemp;
        }
        
        return prev;
    }
};